discoverx/explorer.py

```diff
@@ -0,0 +1,287 @@
+                table_catalog, 
+                table_schema, 
+                table_name
+            FROM {self.columns_table_name}
+            WHERE 
+                table_schema != "information_schema" 
+                {catalog_sql if catalogs != "*" else ""}
+                {schema_sql if schemas != "*" else ""}
+                {table_sql if tables != "*" else ""}
+                {columns_sql if columns else ""}
+        )
+
+        SELECT
+            info_schema.table_catalog, 
+            info_schema.table_schema, 
+            info_schema.table_name, 
+            collect_list(struct(column_name, data_type, partition_index)) as table_columns
+        FROM {self.columns_table_name} info_schema
+        INNER JOIN tb_list ON (
+            info_schema.table_catalog <=> tb_list.table_catalog AND
+            info_schema.table_schema = tb_list.table_schema AND
+            info_schema.table_name = tb_list.table_name)
+        GROUP BY info_schema.table_catalog, info_schema.table_schema, info_schema.table_name
+        """
+
+        return helper.strip_margin(sql)
+
+
+class DataExplorer:
+    from_components_expr = r"^(([0-9a-zA-Z_\*]+)\.([0-9a-zA-Z_\*]+)\.([0-9a-zA-Z_\*]+))$"
+
+    def __init__(self, from_tables, spark: SparkSession, info_fetcher: InfoFetcher) -> None:
+        self._from_tables = from_tables
+        self._catalogs, self._schemas, self._tables = DataExplorer.validate_from_components(from_tables)
+        self._spark = spark
+        self._info_fetcher = info_fetcher
+        self._having_columns = []
+        self._sql_query_template = None
+        self._max_concurrency = 10
+
+    @staticmethod
+    def validate_from_components(from_tables: str):
+        """Extracts the catalog, schema and table name from the from_table string"""
+        matches = re.findall(DataExplorer.from_components_expr, from_tables)
+        if len(matches) == 1 and len(matches[0]) == 4:
+            return (matches[0][1], matches[0][2], matches[0][3])
+        else:
+            raise ValueError(
+                f"Invalid from_tables statement '{from_tables}'. Should be a string in format 'table_catalog.table_schema.table_name'. You can use '*' as wildcard."
+            )
+
+    def __deepcopy__(self, memo):
+        new_obj = type(self).__new__(self.__class__)
+        new_obj.__dict__.update(self.__dict__)
+        new_obj._catalogs = copy.deepcopy(self._catalogs)
+        new_obj._schemas = copy.deepcopy(self._schemas)
+        new_obj._tables = copy.deepcopy(self._tables)
+        new_obj._having_columns = copy.deepcopy(self._having_columns)
+        new_obj._sql_query_template = copy.deepcopy(self._sql_query_template)
+        new_obj._max_concurrency = copy.deepcopy(self._max_concurrency)
+
+        new_obj._spark = self._spark
+        new_obj._info_fetcher = self._info_fetcher
+
+        return new_obj
+
+    def having_columns(self, *columns) -> "DataExplorer":
+        new_obj = copy.deepcopy(self)
```

Comment:
Why do we need to deepcopy? Wouldn't it also be possible to just return the changed `self`?

Reply:
Yes, but that would not be a "functional" pattern, because it would change the state of the original object.
I would rather have a functional pattern like the Spark API.

Reply:
Makes sense!